server/src/test/java/org/elasticsearch/index/replication/RetentionLeasesReplicationTests.java

```diff
@@ -0,0 +1,151 @@
+                listener.onResponse(new SyncRetentionLeasesResponse(new RetentionLeaseSyncAction.Request(shardId, leases)));
+            }
+        }) {
+            group.startAll();
+            int numLeases = between(1, 10);
+            List<RetentionLeaseSyncAction.Request> requests = new ArrayList<>();
+            for (int i = 0; i < numLeases; i++) {
+                PlainActionFuture<ReplicationResponse> future = new PlainActionFuture<>();
+                group.addRetentionLease(Integer.toString(i), randomNonNegativeLong(), "test-" + i, future);
+                requests.add(((SyncRetentionLeasesResponse) future.actionGet()).syncRequest);
+            }
+            RetentionLeases leasesOnPrimary = group.getPrimary().getRetentionLeases();
+            for (IndexShard replica : group.getReplicas()) {
+                Randomness.shuffle(requests);
+                requests.forEach(request -> group.executeRetentionLeasesSyncRequestOnReplica(request, replica));
+                assertThat(replica.getRetentionLeases(), equalTo(leasesOnPrimary));
+            }
+        }
+    }
+
+    public void testSyncRetentionLeasesWithPrimaryPromotion() throws Exception {
+        Settings settings = Settings.builder().put(IndexSettings.INDEX_SOFT_DELETES_SETTING.getKey(), true).build();
+        int numberOfReplicas = between(2, 4);
+        IndexMetaData indexMetaData = buildIndexMetaData(numberOfReplicas, settings, indexMapping);
+        try (ReplicationGroup group = new ReplicationGroup(indexMetaData) {
+            @Override
+            protected void syncRetentionLeases(ShardId shardId, RetentionLeases leases, ActionListener<ReplicationResponse> listener) {
+                listener.onResponse(new SyncRetentionLeasesResponse(new RetentionLeaseSyncAction.Request(shardId, leases)));
+            }
+        }) {
+            group.startAll();
+            int numLeases = between(1, 100);
+            IndexShard newPrimary = randomFrom(group.getReplicas());
+            RetentionLeases latestRetentionLeasesOnNewPrimary = RetentionLeases.EMPTY;
+            for (int i = 0; i < numLeases; i++) {
+                PlainActionFuture<ReplicationResponse> addLeaseFuture = new PlainActionFuture<>();
+                group.addRetentionLease(Integer.toString(i), randomNonNegativeLong(), "test-" + i, addLeaseFuture);
+                RetentionLeaseSyncAction.Request request = ((SyncRetentionLeasesResponse) addLeaseFuture.actionGet()).syncRequest;
+                for (IndexShard replica : randomSubsetOf(group.getReplicas())) {
+                    group.executeRetentionLeasesSyncRequestOnReplica(request, replica);
+                    if (newPrimary == replica) {
+                        latestRetentionLeasesOnNewPrimary = request.getRetentionLeases();
+                    }
+                }
+            }
+            group.promoteReplicaToPrimary(newPrimary).get();
```

Comment:
Discuss: should we align the retention-leases when a new primary is promoted?

Reply:
Under what circumstances can the new primary not hold an up-to-date set of leases already? It might perhaps be missing some renewals but I think that's ok.

Reply:
We are adding two new leases to the old primary: L1 and L2. L1 was synced to replica r1; L2 was synced to r2, but the old primary crashed before two leases are properly synced to all two replicas. If any replica is promoted, then the retention leases between copies are not aligned.

Reply:
We sync by copying _all_ the leases from the primary to its replicas, so I don't follow how r2 could receive L2 without also receiving L1 (assuming they were added in this order).

However, I think I do see a potential problem:
- primary A shares a lease with one replica B, but not to another replica C
- A crashes
- C discards some history that the lease would have retained
- B is promoted to primary and shares its lease with C
- C cannot accept this lease since it has already discarded this history

I think we can prevent this, with peer-recovery retention leases, by insisting that leases do not "go backwards", i.e. they only retain history that is already being retained. This would mean that C could not discard the history in the situation above because it must already hold a different lease that retains that history.